os/board/rtl8721csm/src/component/common/api/wifi_interactive_mode.c

```diff
@@ -728,43 +733,68 @@ int8_t cmd_wifi_on(WiFi_InterFace_ID_t interface_id)
 {
 	int ret;
 
-	/* Kill init thread after all init tasks done */
-	ret = wifi_on(RTW_MODE_STA);
-	if (ret != RTW_SUCCESS) {
-		ndbg("\n\rrtk_wifi_start failed\n");
-		return RTK_STATUS_ERROR;
+	if (rltk_wlan_running(WLAN0_IDX)) {
+		if (wifi_set_mode(RTW_MODE_STA) < 0) {
+			ndbg("\n\rERROR: Wifi Set Mode to STA failed!\n");
+			return RTW_ERROR;
+		}
+	} else {
+		/* Kill init thread after all init tasks done */
+		ret = wifi_on(RTW_MODE_STA);
+		if (ret != RTW_SUCCESS) {
+			ndbg("\n\rrtk_wifi_start failed\n");
+			return RTK_STATUS_ERROR;
+		}
+		nvdbg("\r\n===============>>wifi_on success!!\r\n");
 	}
-	nvdbg("\r\n===============>>wifi_on success!!\r\n");
 
 #if RTW_AUTO_RECONNECT
 	//setup reconnection flag
 	wifi_set_autoreconnect(1);
 #endif
 
-	{
-		rtw_wifi_setting_t setting;
-
-		wifi_get_setting(WLAN0_NAME, &setting);
-		wifi_show_setting(WLAN0_NAME, &setting);
-
-	}
+	rtw_wifi_setting_t setting;
+	wifi_get_setting(WLAN0_NAME, &setting);
+	wifi_show_setting(WLAN0_NAME, &setting);
 
 	nvdbg("\r\n===============>>Finish wifi_on!!\r\n");
 	return RTK_STATUS_SUCCESS;
 }
 
 int8_t cmd_wifi_off(void)
 {
+	int mode;
+
 #if CONFIG_WEBSERVER
 	stop_web_server();
 #endif
 #if CONFIG_ENABLE_P2P
 	cmd_wifi_p2p_stop(0, NULL);
 #else
+	if ((!rltk_wlan_running(WLAN0_IDX)) &&
+		(!rltk_wlan_running(WLAN1_IDX))) {
+		RTW_API_INFO("\n\rWIFI is not running");
+		return 0;
+	}
+	wext_get_mode(WLAN0_NAME, &mode);
+	if (mode == IW_MODE_MASTER) {
+		RTW_API_INFO("\n\rWIFI Mode Change: AP, disable beacon\r\n");
+		wext_set_mode(WLAN0_NAME, IW_MODE_INFRA);
+		rtw_msleep_os(50);
+	} else if (mode == IW_MODE_INFRA) {
+		RTW_API_INFO("\n\rWIFI Mode Change: STA, disconnecting\r\n");

```

Comment:
The log for Wifi disconnecting is printed here. But, disconnect might not be done if the below "if" condition fails. So, can we print this log inside the if statement?